Fix some pod-page ordering nits

richsalz opened the pull request:
The find-doc-nits script only looked for EXAMPLES, not EXAMPLE.
Fix the pattern and then fix the errors that resulted.

InfoHunter commented:
Interesting. For the headings, should they all be `EXAMPLES` even if there is only one example in the section?

levitte commented:
It's a section name and it should be the same accross the board.
perlpodstyle mentions EXAMPLES, Linux' man-pages(7) mentions EXAMPLE, BSD's mdoc(7) mentions EXAMPLES. 

2 out of 3 makes EXAMPLES the winner? Also, I think we have used the plural form often than the singular...

richsalz commented:
Okay, added a commit to rename all sections from EXAMPLE to EXAMPLES and updated find-doc-nits to complain.  And merged multiple consecutive EXAMPLE sections into a single section.

Probably worth backporting this.

mspncp commented on util/find-doc-nits:
```diff
@@ -178,15 +178,15 @@ sub check()
 
+    # Check ordering of some sections in man3
+    if ( $filename =~ m|man3/| ) {
+        &section_follows($id, $contents, "RETURN VALUES", "EXAMPLES");
```
The function name is very confusing, because "follows" reverses the order:  I'm tempted to read it as

```
    section_follows('A', 'B')   =>  Section 'A'  follows section 'B'   =>   First comes section 'B', then section 'A' 
```
My alternative proposal would be to rename `check_section_location()` (which is admittedly not very expressive) to `check_section_order()` .

richsalz replied:
Name changed to ```check_section_location``` and fixup commit pushed.
(I avoided the temptation to name it ```section_preceeds``` :)

mspncp commented:
> It's a section name and it should be the same accross the board.

Agreed. And that is the reason why it should alway be plural.

BTW: `RETURN VALUES` is always plural (I checked it), even if there is only one function documented (this I didn't check; but I'm sure there is an example ;-).

mspncp commented:
> Probably worth backporting this.

I added [master] and [1.1.1] labels. Do you agree @levitte?

levitte commented:
How cleanly does it cherry pick to 1.1.1?  Mind you, I don't mind a backport, so I agree with the sentiment, but it may be that it needs a separate PR

mspncp commented:
> How cleanly does it cherry pick to 1.1.1?

Isn't that the usual disclaimer? "...provided that it cherry-picks cleanly."  ;-)

mspncp commented on util/find-doc-nits:
```diff
@@ -151,17 +151,17 @@ sub name_synopsis()
 }
 
-# Check if SECTION is located before BEFORE
+# Check if SECTION ($3) is located before BEFORE ($4)
```
How can something be located before "BEFORE"?  Shouldn't it be located before "AFTER"?

_Never mind, forget my comment. Just picking nits on the nit-picker. 😜_

richsalz commented:
Added a commit that enforces WARNINGS (over WARNING), similar to the EXAMPLES/EXAMPLE
nit.

richsalz commented:
@levitte and @mspncp I added a commit (not noticing that you had approved it) so this needs re-approval.

If there are issues about backporting, let me know and I can create a backport PR.

mspncp commented:
The fixup did not commute (because of the renaming), so I had to resolve some conflicts. Unfortunately , the conflict happened after I already rebased on master. I should have repeated the interactive merge without moving the base on master, then I could have used the [force-push diff](https://github.com/openssl/openssl/compare/20020bdd8aabe4a7e27b305535ce2a6c7804ba53..7aec8f0a32875bfd705b713ad73f4ef61ac8f113) to prove that no code was changed. Unfortunately, I forgot to do so.

So please recheck the three changesets, @richsalz.

mspncp commented:
Well, the force-push diff shows at least that only unrelated code was changed :grin:

richsalz commented:
LGTM, @mspncp .

mspncp commented:
Richard, the cherry-pick to 1.1.1 of cda774223d fails. Most of the conflicts are deletions of pod files, which didn't exist yet, but there are also textual conflicts. Better you take a look and raise a separate pr.

```
On branch OpenSSL_1_1_1-stable
Your branch is up to date with 'origin/OpenSSL_1_1_1-stable'.

You are currently cherry-picking commit cda774223d.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Changes to be committed:

	modified:   doc/man1/engine.pod
	modified:   doc/man1/errstr.pod
	modified:   doc/man1/pkeyparam.pod
	modified:   doc/man3/ASYNC_start_job.pod
	modified:   doc/man3/BIO_find_type.pod
	modified:   doc/man3/BIO_new.pod
	modified:   doc/man3/BIO_s_accept.pod
	modified:   doc/man3/BIO_s_bio.pod
	modified:   doc/man3/BIO_s_connect.pod
	modified:   doc/man3/BIO_s_fd.pod
	modified:   doc/man3/CRYPTO_THREAD_run_once.pod
	modified:   doc/man3/EVP_DigestInit.pod
	modified:   doc/man3/EVP_PKEY_CTX_set_hkdf_md.pod
	modified:   doc/man3/EVP_PKEY_CTX_set_tls1_prf_md.pod
	modified:   doc/man3/EVP_PKEY_decrypt.pod
	modified:   doc/man3/EVP_PKEY_derive.pod
	modified:   doc/man3/EVP_PKEY_encrypt.pod
	modified:   doc/man3/EVP_PKEY_sign.pod
	modified:   doc/man3/EVP_PKEY_verify.pod
	modified:   doc/man3/EVP_PKEY_verify_recover.pod
	modified:   doc/man3/OCSP_REQUEST_new.pod
	modified:   doc/man3/PKCS12_newpass.pod
	modified:   doc/man3/SSL_CTX_config.pod
	modified:   doc/man3/SSL_CTX_dane_enable.pod
	modified:   doc/man3/SSL_set1_host.pod
	modified:   doc/man3/X509_VERIFY_PARAM_set_flags.pod
	modified:   doc/man7/Ed25519.pod
	modified:   doc/man7/SM2.pod
	modified:   doc/man7/X25519.pod
	modified:   doc/man7/bio.pod
	modified:   doc/man7/scrypt.pod

Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)

	both modified:   doc/man3/BIO_set_callback.pod
	deleted by us:   doc/man3/EVP_MAC.pod
	deleted by us:   doc/man3/OSSL_CMP_ITAV_set0.pod
	deleted by us:   doc/man3/OSSL_CRMF_pbmp_new.pod
	deleted by us:   doc/man3/OSSL_PARAM_construct_from_text.pod
	both modified:   doc/man3/SSL_CTX_get0_param.pod
	deleted by us:   doc/man7/EVP_KDF_HKDF.pod
	deleted by us:   doc/man7/EVP_KDF_SS.pod
	deleted by us:   doc/man7/EVP_KDF_SSHKDF.pod
	deleted by us:   doc/man7/EVP_KDF_TLS1_PRF.pod
	deleted by us:   doc/man7/EVP_KDF_X942.pod
	deleted by us:   doc/man7/EVP_KDF_X963.pod
	both modified:   util/find-doc-nits
```

mspncp commented:
> Richard, ...

Sorry, I meant Rich (@richsalz). I better go to bed now...

richsalz commented:
Backport in #9632

mspncp commented:
> The fixup did not commute (because of the renaming), so I had to resolve some conflicts. 
> ...
> So please recheck the three changesets, @richsalz.

I should have known you better, Rich Salz. Then I wouldn't have wasted my time to properly rebase the fixup and keep your three commits. ;-) (Well, actually I _did_ know it, but I never squash other peoples commits without asking, if they are not marked explicitly.)

Since you carelessly squashed your commits for the backport #9632, I think this pull request should be squashed, too.

levitte commented:
I'm a little lost with what has gone on here since you made your last approval, @mspncp.  Do you approve of what's in here now?  In that case, I can merge.

levitte commented:
> Sorry if I was unclear.

Not a problem.  It was actually github display that was unclear...  I now understand that @richsalz's commits that are shown above your force-push were a result of that force-push.  I really shouldn't work before my second coffee :wink:

Anyway, an extra approval is no harm...

levitte commented:
Oh!  I didn't see this has been merged already.  Is there a reason this PR is kept open?

mspncp commented:
The reason is that I actually was trying to cherry-pick the commits to 1.1.1 and then got stuck with conflicts. So I asked @richsalz to open a separate pull request and removed the [1.1.1] label. That's when I forgot to close it, sorry. I really shouldn't work so late at night either ;-)

mspncp commented:
> Since you carelessly squashed your commits for the backport #9632, I think this pull request should be squashed, too.

Obviously I had forgotten about it this morning :-/ Probably did too many things at the same time. Never mind.

levitte commented:
I guess it's time to close, then, since it's in...

richsalz commented:
>carelessly squashed

It wasn't careless.  Like a bug on the sidewalk, I mindfully brought down the boot :)  In general, unless there are strong extenuating circumstances, I prefer to squash everything in a PR.  But I leave what happens up to the committer who does the merge.

FYI, this is how I created the backport PR:
```
git co fix-doc-ordering
git diff master > /tmp/diff
git co OpenSSL_1_1_1-stable
git co -b backport-fix-doc-ordering
patch -p1 < /tmp/diff
# whenever it can't find a file, hit return for the filename and then return a second time to skip
# the patch
./config
make doc-nits
# Fix up the one older file that needed to be fixed
git commit .
```
